feat: user management API

victor-borges opened the pull request.

NikiforovAll commented:
I think we should start adding unit tests for Sdks, is it something that you are interested in doing? 

ref: https://github.com/richardszalay/mockhttp

NikiforovAll commented on src/Keycloak.AuthServices.Sdk/Admin/Constants/KeycloakClientApiConstants.cs:
```diff
@@ -25,4 +25,12 @@ internal static class KeycloakClientApiConstants
+
+    internal const string GetUsers = $"{GetRealm}/users";
+
+    internal const string CreateUser = $"{GetUsers}";
```
Drop redundant string interpolation

victor-borges replied:
So do we reuse the `GetUsers` constant?

NikiforovAll replied:
I think so, also, not sure if reusing the constants is good idea, you might as well refactor all constants to plain strings, reasonable duplication without overcomplicating

NikiforovAll replied:
Maybe, it was a bad choice on my side :)

victor-borges replied:
For the discussion, there are 2 different endpoints that share that route, Create User (`POST /{realm}/users`) and Get Users (`GET /{realm}/users`)

victor-borges replied:
> reasonable duplication without overcomplicating

Fully agreed

NikiforovAll replied:
I'm fine with the duplication

NikiforovAll commented on src/Keycloak.AuthServices.Sdk/Admin/Models/User.cs:
```diff
@@ -0,0 +1,70 @@
+/// </summary>
+public class User
+{
+    public Dictionary<string, string>? Access { get; init; }
+    public Dictionary<string, string>? Attributes { get; init; }
+    public UserConsent[]? ClientConsents { get; init; }
+    public Dictionary<string, string>? ClientRoles { get; init; }
+    public long? CreatedTimestamp { get; init; }
+    public Credential[]? Credentials { get; init; }
+    public string[]? DisableableCredentialTypes { get; init; }
+    public string? Email { get; init; }
+    public bool? EmailVerified { get; init; }
+    public bool? Enabled { get; init; }
+    public FederatedIdentity? FederatedIdentities { get; init; }
+    public string? FederationLink { get; init; }
+    public string? FirstName { get; init; }
+    public string[]? Groups { get; init; }
+    public string? Id { get; init; }
+    public string? LastName { get; init; }
+    public int? NotBefore { get; init; }
+    public string? Origin { get; init; }
+    public Dictionary<string, string>? RealmRoles { get; init; }
+    public Dictionary<string, string>? RequiredActions { get; init; }
+    public string? Self { get; init; }
+    public string? ServiceAccountClientId { get; init; }
+    public string? Username { get; init; }
```
Are there required fields? A lot of nullables, I'm wondering if we could help the consumers with more precise model

victor-borges replied:
At least in the docs they are really all optional
https://www.keycloak.org/docs-api/20.0.2/rest-api/index.html#_userrepresentation

NikiforovAll replied:
OK, looks good

NikiforovAll commented on src/Keycloak.AuthServices.Sdk/Admin/Models/User.cs:
```diff
@@ -0,0 +1,70 @@
+/// <summary>
+/// Credential representation.
+/// </summary>
+public class Credential
```
I guess we should move a class to a separate file, might be a good idea to organize models in sub-folders based on functionality

victor-borges replied:
I was thinking of that as well, will do!

victor-borges commented:
> I think we should start adding unit tests for Sdks, is it something that you are interested in doing?
> 
> ref: https://github.com/richardszalay/mockhttp

Yes!

NikiforovAll reviewed (commented):
That is a good start, thanks

NikiforovAll commented:
Things to consider:

* Update readme 
* Add tests for other API Clients

Please move the PR from Draft once you consider the development finished. Thank for working on this

NikiforovAll commented on test/Keycloak.AuthServices.Sdk.Tests/Admin/KeycloakRealmClientTests.cs:
```diff
@@ -0,0 +1,45 @@
+    [Fact]
+    public async Task GetRealmShouldThrowNotFoundApiExceptionWhenRealmDoesNotExist()
+    {
+        this.handler.Expect(HttpMethod.Get, $"{BaseAddress}/admin/realms/inexistent")
```
I guess `inexistent` is a Spanish word 😅, could we use something like `nonexistent`

NikiforovAll commented on test/Keycloak.AuthServices.Sdk.Tests/Admin/KeycloakRealmClientTests.cs:
```diff
@@ -0,0 +1,45 @@
+            .Respond(HttpStatusCode.NotFound);
+
+        var exception = await Assert.ThrowsAsync<ApiException>(
+            async () => await this.keycloakRealmClient.GetRealm("inexistent"));
```
Just wondering, do we need `async/await`, does it improve stack trace or something? We can just return `Task` I guess

NikiforovAll commented on test/Directory.Build.props:
```diff
@@ -0,0 +1,23 @@
+<Project>
```
Not related to the content of the file 

please use tests as folder name

NikiforovAll reviewed (commented):
Good job! Let's finalize it

NikiforovAll commented:
LGTM, I plan to run it locally and we can integrate the changes into main. Have you tried using the SDK Client against running Keycloak?

victor-borges commented:
> LGTM, I plan to run it locally and we can integrate the changes into main. Have you tried using the SDK Client against running Keycloak?

Yes, I did now. I changed the AuthGettingStarted project like this (also changed the realm settings to match mine):

```csharp
app.MapPost("/user", async (IKeycloakUserClient keycloakUserClient) =>
{
    var user = new User {Username = "email@domain.com"};
    await keycloakUserClient.CreateUser("master", user);
    return Results.Ok(user);
});
```

Worked as expected, user token is propagated to the request, but now I realized that some of the status codes in the unit tests are kinda wrong, it returns 201 Created for a successful operation.

NikiforovAll commented:
Could you please check the failed job?

victor-borges commented:
The `dotnet test` command that Cake runs is expecting the `.dll` files to be in the `tests/Keycloak.AuthServices.Sdk.Tests/bin/Release/net6.0/`, but the previous `Build` cake task puts them in the `Artefacts` folder. I'm not sure how to point it to there.

victor-borges commented:
Got the tests to run like this:

```diff
Task("Test")
    .Description("Runs unit tests and outputs test results to the artefacts directory.")
-   .DoesForEach(GetFiles("./tests/**/*.csproj"), project =>
+   .DoesForEach(GetFiles($"{artefactsDirectory}/*Tests.dll"), project =>
```

but then there's another problem with the coverlet DLLs, that are not copied to the Artefacts 
folder (also, should it be renamed to Artifacts)?

```
> dotnet cake --target=Test   

========================================
Test
========================================
The following arguments have been ignored : "--configuration Release --no-build --no-restore"
Microsoft (R) Test Execution Command Line Tool Version 17.3.1 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
Data collection : Unable to find a datacollector with friendly name 'XPlat Code Coverage'.
Data collection : Could not find data collector 'XPlat Code Coverage'
WARNING: Overwriting results file: /home/victor-borges/Source/keycloak-authorization-services-dotnet/Artefacts/Keycloak.AuthServices.Sdk.Tests.trx
Results File: /home/victor-borges/Source/keycloak-authorization-services-dotnet/Artefacts/Keycloak.AuthServices.Sdk.Tests.trx
Html test results file : /home/victor-borges/Source/keycloak-authorization-services-dotnet/Artefacts/Keycloak.AuthServices.Sdk.Tests.html

Passed!  - Failed:     0, Passed:     4, Skipped:     0, Total:     4, Duration: 13 ms - /home/victor-borges/Source/keycloak-authorization-services-dotnet/Artefacts/Keycloak.AuthServices.Sdk.Tests.dll (net6.0)

Task                          Duration            
--------------------------------------------------
Test                          00:00:01.6253500    
--------------------------------------------------
Total:                        00:00:01.6253500
```

NikiforovAll commented:
> The `dotnet test` command that Cake runs is expecting the `.dll` files to be in the `tests/Keycloak.AuthServices.Sdk.Tests/bin/Release/net6.0/`, but the previous `Build` cake task puts them in the `Artefacts` folder. I'm not sure how to point it to there.

Thanks, I will take a look

NikiforovAll commented:
Hi @victor-borges ,

I've created PR in your fork, let's see how it goes

victor-borges commented:
> I've created PR in your fork, let's see how it goes

Merged it, just waiting for approval now

NikiforovAll commented:
Congratulates on the first closed MR, thank you for the contribution, I appreciate it! Please feel free to contribute more, you already have the valuable experience

NikiforovAll commented:
ref: https://github.com/NikiforovAll/keycloak-authorization-services-dotnet/releases/tag/1.4.0

victor-borges commented:
Thanks you so much for the support!
